src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,746 @@
+ *           advised that {@linkplain java.lang.ref##reachability reachable} stable values
+ *           will hold their set content perpetually.
+ *           A {@linkplain StableValue} that has a type parameter {@code T} that is an array
+ *           type (of arbitrary rank) will only allow the JVM to treat the <em>array reference</em>
+ *           as a stable value but <em>not its components</em>. Clients can instead use
+ *           {@linkplain #list(int, IntFunction) a stable list} of arbitrary depth, which
+ *           provides stable components. More generally, a stable value can hold other
+ *           stable values of arbitrary depth and still provide transitive constantness.
+ *
+ * @param <T> type of the content
+ *
+ * @since 25
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.STABLE_VALUES)
+public sealed interface StableValue<T>
+        permits StableValueImpl {
+
+    // Principal methods
+
+    /**
+     * Tries to set the content of this StableValue to the provided {@code content}.
+     * <p>
+     * When this method returns, the content of this StableValue is always set.
+     *
+     * @return {@code true} if the content of this StableValue was set to the
+     *         provided {@code content}, {@code false} otherwise
+     * @param content to set
+     * @throws IllegalStateException if this method is invoked directly by a supplier
+     *         provided to the {@link #orElseSet(Supplier)} method.
+     */
+    boolean trySet(T content);
+
+    /**
+     * {@return the content if set, otherwise, returns the provided {@code other} value}
+     *
+     * @param other to return if the content is not set
+     */
+    T orElse(T other);
+
+    /**
+     * {@return the content if set, otherwise, throws {@code NoSuchElementException}}
+     *
+     * @throws NoSuchElementException if no content is set
+     */
+    T orElseThrow();
+
+    /**
+     * {@return {@code true} if the content is set, {@code false} otherwise}
+     */
+    boolean isSet();
+
+    /**
+     * {@return the content; if unset, first attempts to compute and set the
+     *          content using the provided {@code supplier}}
+     * <p>
+     * The provided {@code supplier} is guaranteed to be invoked at most once if it
+     * completes without throwing an exception.
+     * <p>
+     * If the supplier throws an (unchecked) exception, the exception is rethrown, and no
+     * content is set. The most common usage is to construct a new object serving
+     * as a lazily computed value or memoized result, as in:
+     *
+     * {@snippet lang=java:
+     * Value witness = stable.orElseSet(Value::new);
+     * }
+     * <p>
+     * When this method returns successfully, the content is always set.
+     * <p>
+     * The provided {@code supplier} will only be invoked once even if invoked from
+     * several threads unless the {@code supplier} throws an exception.
```

Comment:
Is this the supplier instance or any supplier? Just wondering about the case where two threads are invoke it with different instances (for whatever reason), I can't tell from the javadoc.

Reply:
I personally think this is more of an implementation artifact for `StableValue::supplier` and doesn't merit its own API (as no other StableValue API utilizes the object monitor of a StableValue), but maybe we can adjust this in a later preview...

Reply:
I think "throws an exception" is really exception and error. I'm not suggesting this has to be changed now but if you are re-visiting this paragraph then this could be made a bit clearer.